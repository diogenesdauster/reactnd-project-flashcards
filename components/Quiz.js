import React, { Component } from "react";
import PropTypes from "prop-types";
import {
  View,
  Text,
  TouchableOpacity,
  StyleSheet,
  Button,
  Animated
} from "react-native";
import Score from "./Score";
import { white, red, green } from "../utils/colors";

class Quiz extends Component {
  static propTypes = {
    navigation: PropTypes.shape({
      goBack: PropTypes.func.isRequired,
      state: PropTypes.shape({
        params: PropTypes.shape({
          deck: PropTypes.shape({ questions: PropTypes.array.isRequired })
        })
      })
    }).isRequired
  };

  state = {
    cardIndex: 0,
    correctAnswer: 0,
    showAnswer: false
  };
  componentWillMount() {
    this.cardFlip = new Animated.Value(0);
    this.flipValue = 0;
    this.cardFlip.addListener(({ value }) => {
      this.flipValue = value;
    });

    this.flipCardInterpolate = this.cardFlip.interpolate({
      inputRange: [0, 360],
      outputRange: ["0deg", "360deg"]
    });

    this.cardSwipe = new Animated.Value(0);
    this.cardOpacityInterpolate = this.cardSwipe.interpolate({
      inputRange: [-400, 0, 400],
      outputRange: [0, 1, 0]
    });
  }
  componentWillUnmount() {
    this.cardFlip.removeAllListeners();
  }
  _flipCard = () => {
    if (this.flipValue >= 90) {
      Animated.spring(this.cardFlip, {
        toValue: 0,
        friction: 8,
        tension: 10,
        useNativeDriver: true
      }).start();
    } else {
      Animated.spring(this.cardFlip, {
        toValue: 360,
        friction: 8,
        tension: 10,
        useNativeDriver: true
      }).start();
    }
  };
  _swipeCard = () => {
    Animated.sequence([
      Animated.timing(this.cardSwipe, {
        toValue: 400,
        duration: 500,
        useNativeDriver: true
      }),
      Animated.timing(this.cardSwipe, {
        toValue: 0,
        duration: 500,
        useNativeDriver: true
      })
    ]).start();
  };
  _toggleAnswer = () => {
    this._flipCard();
    setTimeout(() => {
      this.setState({ showAnswer: !this.state.showAnswer });
    }, 250);
  };
  _handleButtonClick = isCorrectAnswer => () => {
    let { correctAnswer, cardIndex } = this.state;
    correctAnswer += isCorrectAnswer ? 1 : 0;
    ++cardIndex;
    this._swipeCard();
    setTimeout(() => {
      this.setState({ correctAnswer, cardIndex, showAnswer: false });
    }, 500);
  };
  _renderQuizView = () => {
    const { questions } = this.props.navigation.state.params.deck;
    const { showAnswer, cardIndex } = this.state;
    if (showAnswer) {
      return (
        <View style={{ flex: 1 }}>
          <View>
            <Text>{`${cardIndex + 1} / ${questions.length}`}</Text>
          </View>
          <View style={styles.content}>
            <Text style={styles.answer}>{questions[cardIndex].answer}</Text>
            <Button onPress={this._toggleAnswer} title="View Question" />
          </View>
          <View style={styles.buttonGroup}>
            <TouchableOpacity
              style={[styles.button, { backgroundColor: red }]}
              onPress={this._handleButtonClick(false)}
            >
              <Text style={styles.buttonText}>Incorrect</Text>
            </TouchableOpacity>
            <TouchableOpacity
              style={[styles.button, { backgroundColor: green }]}
              onPress={this._handleButtonClick(true)}
            >
              <Text style={styles.buttonText}>Correct</Text>
            </TouchableOpacity>
          </View>
        </View>
      );
    }
    return (
      <View style={{ flex: 1 }}>
        <View>
          <Text>{`${cardIndex + 1} / ${questions.length}`}</Text>
        </View>
        <View style={styles.content}>
          <Text style={styles.question}>{questions[cardIndex].question}</Text>
          <Button onPress={this._toggleAnswer} title="View Answer" />
        </View>
      </View>
    );
  };
  _startAgain = () => {
    this.setState({
      cardIndex: 0,
      correctAnswer: 0,
      showAnswer: false
    });
  };
  _goBackToDeck = () => {
    this.props.navigation.goBack();
  };
  render() {
    const animatedStyle = {
      transform: [
        { rotateY: this.flipCardInterpolate },
        { translateX: this.cardSwipe }
      ],
      opacity: this.cardOpacityInterpolate
    };
    const { questions } = this.props.navigation.state.params.deck;
    const { cardIndex, correctAnswer } = this.state;
    return (
      <Animated.View style={[styles.container, animatedStyle]}>
        {cardIndex < questions.length ? (
          this._renderQuizView()
        ) : (
          <Score
            score={Number((correctAnswer / questions.length) * 100).toFixed(2)}
            startAgain={this._startAgain}
            goBackToDeck={this._goBackToDeck}
          />
        )}
      </Animated.View>
    );
  }
}

export default Quiz;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    margin: 20,
    padding: 20,
    backgroundColor: white,
    shadowRadius: 3,
    shadowOpacity: 0.8,
    shadowColor: "rgba(0, 0, 0, 0.24)",
    shadowOffset: {
      width: 0,
      height: 3
    }
  },
  content: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center"
  },
  buttonGroup: {
    height: 50,
    flexDirection: "row",
    justifyContent: "space-between"
  },
  question: {
    fontSize: 32,
    textAlign: "center",
    marginBottom: 10
  },
  answer: {
    fontSize: 28,
    textAlign: "center",
    marginBottom: 10
  },
  button: {
    padding: 10,
    width: 100,
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    marginLeft: 5,
    marginRight: 5,
    shadowRadius: 3,
    shadowOpacity: 0.8,
    shadowColor: "rgba(0, 0, 0, 0.24)",
    shadowOffset: {
      width: 0,
      height: 3
    }
  },
  buttonText: {
    color: white,
    fontSize: 18
  }
});
