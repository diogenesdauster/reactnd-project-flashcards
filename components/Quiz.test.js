import "react-native";
import React from "react";
import Quiz from "./Quiz";
// Note: test renderer must be required after react-native.
import renderer from "react-test-renderer";

// static propTypes = {
//     navigation: PropTypes.shape({
//       goBack: PropTypes.func.isRequired,
//       state: PropTypes.shape({
//         params: PropTypes.shape({
//           deck: PropTypes.shape({ questions: PropTypes.array.isRequired })
//         })
//       })
//     }).isRequired
//   };

it("renders Quiz correctly", () => {
  const mockNavigation = {
    goBack: jest.fn(),
    state: {
      params: {
        deck: {
          questions: [
            {
              question: "What is React?",
              answer: "A library for managing user interfaces"
            },
            {
              question: "Where do you make Ajax requests in React?",
              answer: "The componentDidMount lifecycle event"
            }
          ]
        }
      }
    }
  };
  const tree = renderer.create(<Quiz navigation={mockNavigation} />).toJSON();
  expect(tree).toMatchSnapshot();
});
