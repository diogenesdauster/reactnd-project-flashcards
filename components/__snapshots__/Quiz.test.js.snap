// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Quiz correctly 1`] = `
<View
  collapsable={undefined}
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "margin": 20,
      "opacity": 1,
      "padding": 20,
      "shadowColor": "rgba(0, 0, 0, 0.24)",
      "shadowOffset": Object {
        "height": 3,
        "width": 0,
      },
      "shadowOpacity": 0.8,
      "shadowRadius": 3,
      "transform": Array [
        Object {
          "rotateY": "0deg",
        },
        Object {
          "translateX": 0,
        },
      ],
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        1 / 2
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 32,
            "marginBottom": 10,
            "textAlign": "center",
          }
        }
      >
        What is React?
      </Text>
      <View
        accessibilityComponentType="button"
        accessibilityLabel={undefined}
        accessibilityTraits={
          Array [
            "button",
          ]
        }
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={undefined}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            View Answer
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
